File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    Map<Integer,List<Integer>> wire = new HashMap<>();
    
    public int bfs(int i, int a, int b) {
        Queue<Integer> q = new ArrayDeque<>();
        boolean[] visited = new boolean[wire.size()+1];
        q.add(i);
        visited[i] = true;
        int count = 1;
        
        while(!q.isEmpty()) {
            int cur = q.poll();
            for(int x: wire.get(cur)) {
                if(cur == a && x == b) continue;
                if(cur == b && x == a) continue;
                if(!visited[x]) {
                    q.add(x);
                    visited[x] = true;
                    count++;
                }
            }
        }
        return count;
    }
    public int solution(int n, int[][] wires) {
        for(int[] w: wires) {
            wire.computeIfAbsent(w[0],k->new ArrayList<>()).add(w[1]);
            wire.computeIfAbsent(w[1],k->new ArrayList<>()).add(w[0]);
        }
        
        int answer = wire.size();
        for(int[] w: wires) {
            int a = w[0], b = w[1];
            int cnt = bfs(a,a,b);
            int diff = Math.abs(n-cnt-cnt);
            answer = Math.min(answer,diff);
        }
        
        return answer;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.swift
import Foundation

func solution(_ n:Int, _ wires:[[Int]]) -> Int {
    
    var dict:[Int: [Int]] = [:]
    for wire in wires {
        dict[wire[0], default: []].append(wire[1])
        dict[wire[1], default: []].append(wire[0])
    }
    
    func bfs(_ st: Int, _ cut: Int) -> Int {
        var q = [st]
        var head = 0
        var visited = Array(repeating: false, count: wires.count+2)
        visited[st] = true
        var count = 1
        
        while(head < q.count) {
            let cur = q[head]
            head += 1
            
            if let arr = dict[cur] {
                for nxt in arr {
                    if (cur == st && nxt == cut) || (cur==cut && nxt==st) {
                        continue
                    }
                    
                    if(!visited[nxt]) {
                        q.append(nxt)
                        visited[nxt] = true
                        count += 1
                    }
                }   
            }
        }
        return count
    }
    
    var answer = wires.count+2
    for (k,v) in dict {
        for x in v {
            let diff = abs(bfs(k,x) - bfs(x,k))
            answer = min(answer, diff)
        }
    }
    
    return answer
}